skimage/transform/_geometric.py

```diff
@@ -75,31 +77,49 @@ def _center_and_normalize_points(points):
 
     norm_factor = np.sqrt(d) / rms
 
-    part_matrix = norm_factor * np.concatenate(
-        (np.eye(d), -centroid[:, np.newaxis]), axis=1
-    )
-    matrix = np.concatenate(
-        (
-            part_matrix,
-            [
-                [
-                    0,
-                ]
-                * d
-                + [1]
-            ],
-        ),
-        axis=0,
-    )
+    matrix = np.eye(d + 1)
+    matrix[:d, d] = -centroid
+    matrix[:d, :] *= norm_factor
+    return matrix, _apply_homogeneous(matrix, points)
 
-    points_h = np.vstack([points.T, np.ones(n)])
 
-    new_points_h = (matrix @ points_h).T
+def _apply_homogeneous(matrix, points):
+    """Transform (N, D) `points` array with homogeneous (D+1, D+1) `matrix`.
 
-    new_points = new_points_h[:, :d]
-    new_points /= new_points_h[:, d:]
+    Parameters
+    ----------
+    matrix : (D+1, D+1) array_like
+        The transformation matrix to obtain the new points. Note that any
+        object with an `__array__` method [1]_ that returns a matrix with the
+        correct dimensions can be used as input here. This includes all
+        subclasses of :class:`ProjectiveTransform`, for example.
+    points : (N, D) array
+        The coordinates of the image points.
 
-    return matrix, new_points
+    Returns
+    -------
+    new_points : (N, D) array
+        The transformed image points.
+
+    References
+    ----------
+    .. [1]:
+        https://numpy.org/doc/stable/user/basics.interoperability.html#using-arbitrary-objects-in-numpy
+    """
+    points = np.array(points, copy=NP_COPY_IF_NEEDED, ndmin=2)
+    points_h = _as_h(points)
+    new_points_h = points_h @ matrix.T
+    # We divide by the last dimension of the homogeneous
+    # coordinate matrix. In order to avoid division by zero,
+    # we replace exact zeros in this column with a very small number.
+    divs = new_points_h[:, -1]
+    divs = np.where(divs == 0, np.finfo(float).eps, divs)
+    return new_points_h[:, :-1] / divs[:, None]
+
+
+def _as_h(points):
```

Comment:
I know this is a convenience method, but right now I think it makes the code more opaque. The name assumes a bit too much knowledge about knowing the convention of what `h` describes. The code itself isn't actually that complicated. How about `_append_ones_column(...)` or similar?

Reply:
OK - how about `_append_ones_col`?

Reply:
I would prefer `_append_homogeneous_coord` or `append_homogeneous_dim`, because it specifies the intent of the ones col. See my suggestion below.

Reply:
Actually I won't do a GH suggestion because it needs updates in lots of places. So just a code snippet:

```python
def _append_homogeneous_coord(points):
    """Append a column of ones to the right of `points`.

    This creates the representation of the points in the homogeneous coordinate
    space used by homogeneous matrix transforms.

    Parameters
    ----------
    points : array, shape (N, D)
        The input coordinates, where N is the number of points and D is the
        dimension of the coordinate space.

    Returns
    -------
    points_h : array, shape (N, D+1)
        The same points as homogeneous coordinates.
    """
```

Reply:
Done.